src/Elmish.WPF/WpfProgram.fs

```diff
@@ -0,0 +1,130 @@
+﻿namespace Elmish.WPF
+
+open System.Windows
+open Microsoft.Extensions.Logging
+open Microsoft.Extensions.Logging.Abstractions
+open Elmish
+
+
+type WpfProgram<'model, 'msg> =
+  internal {
+    ElmishProgram: Program<unit, 'model, 'msg, Binding<'model, 'msg> list>
+    LoggerFactory: ILoggerFactory
+    /// Only log calls that take at least this many milliseconds. Default 1.
+    BindingPerformanceLogThresholdMs: int
+  }
+
+
+[<RequireQualifiedAccess>]
+module WpfProgram =
+
+
+  let private fromElmishProgram program =
+    { ElmishProgram = program
+      LoggerFactory = NullLoggerFactory.Instance
+      BindingPerformanceLogThresholdMs = 1 }
+
+
+  /// Creates a WpfProgram that does not use commands.
+  let mkSimple
+      (init: unit -> 'model)
+      (update: 'msg  -> 'model -> 'model)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkSimple init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
+
+
+  /// Creates a WpfProgram that uses commands
+  let mkProgram
+      (init: unit -> 'model * Cmd<'msg>)
+      (update: 'msg  -> 'model -> 'model * Cmd<'msg>)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkProgram init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
```

Comment:
By immediately calling `fromElmishProgram`, there is no chance for the user to call `Program` mapping functions from the `Elmish` module.  Is it too much of a burden to have the user call this function?

Reply:
I have attempted to add all relevant `Program` mappings to `WpfProgram`, but I was thinking we could add `mapElmishProgram` function as a fallback, so users can do anything they want. What do you think?

Reply:
For example, I want to call `Elmish.Program.withErrorHandler`.  For this specific case, an alternative is to add another logging category and pipe this data into the logging abstraction.

Another function worth considering is `Elmish.Program.withSubscription`.  Its [current implementation is equivalent to returning the command from `init`](https://github.com/elmish/elmish/issues/183#issuecomment-526826760), but [its behavior might change in v4 of Elmish](https://github.com/elmish/elmish/issues/194#issue-516190007).

A user might also have a reason to call `withTermination`, though I don't expect to myself.

Reply:
But the reason I haven't added it yet, is that I see only a slight benefit, and several drawbacks. The drawbacks stem from the fact that there is no current `Program` functions that are relevant for the user, and several that may be misleading:

The following are only intended for library authors:
* `map`
* `mapErrorHandler` (**Note to self:** This should be logged in this PR!)
* `setState`
* `view`
* `withSetState`
* `withSyncDispatch`

The following are already present in `WpfProgram` (some adjusted for Elmish.WPF):
* `mkSimple`
* `mkProgram`
* `withSubscription`

The following should not be used in Elmish.WPF:
* `run`
* `runWith`
* `withErrorHandler` (would break logging set up using `WpfProgram.withLogger`)
* `withConsoleTrace` (superfluous since it's controlled using `WpfProgram.withLogger`)
* `withTrace` (would break logging set up using `WpfProgram.withLogger`)

In other words, giving Elmish.WPF users the ability to map `Program` would at best be useless and at worst cause problems. In a sense, the `Program` state as used in `Elmish.WPF` is an internal implementation detail and contains invariants that must be guarded.

The only benefit in adding `mapElmishProgram` is that if new `Program` functions are added to Elmish in the future, then users can use them without Elmish.WPF needing to be updated. Given the nature of Elmish, I do not think it's very likely that there'll be frequent additions of `Program` functions that are useful for Elmish.WPF users, and if they are added, I have no problem simply adding wrappers to `WpfProgram`, like I have done with `withSubscription`.

In case you disagree with my listing above and think that some functions may be useful to Elmish.WPF users; sure, let's discuss that separately. It doesn't change the fundamental reasoning behind my position; we can just add (possibly adapted) wrappers for the useful functionality.

Reply:
> For example, I want to call `Elmish.Program.withErrorHandler`. For this specific case, an alternative is to add another logging category and pipe this data into the logging abstraction.

I just pushed a commit doing that.

> Another function worth considering is `Elmish.Program.withSubscription`. Its [current implementation is equivalent to returning the command from `init`](https://github.com/elmish/elmish/issues/183#issuecomment-526826760), but [its behavior might change in v4 of Elmish](https://github.com/elmish/elmish/issues/194#issue-516190007).

That was already present: `WpfProgram.withSubscription`

> A user might also have a reason to call `withTermination`, though I don't expect to myself.

I can't see that function on my end.

Reply:
`withTermination` was added in [this commit](https://github.com/elmish/elmish/commit/1fe85cf1f1b28e1070b3dd661a0e2f591bb7eb2e).  I don't think it is released yet.

Reply:
I am not convinced that routing Elmish logging through Elmish.WPF use of Microsoft's logging abstraction is a good idea.  Currently the user directly interacts with Elmish logging via functions like `withErrorHandler`, `withConsoleTrace` and `withTrace`.  This is very clear.  The proposed alternative is to route these log events through Elmish.WPF and its use of Microsoft's logging abstraction under the category name `Elmish.WPF.Update`.  Even if a better name is selected for this category, this approach is less clear.

> `withConsoleTrace` (superfluous since it's controlled using `WpfProgram.withLogger`)

I wish it were that simple, but as I pointed out in https://github.com/elmish/elmish/issues/196, calling `withConsoleTrace` is not the same as calling `withTrace` and writing to the console.

Reply:
> Even if a better name is selected for this category, this approach is less clear.

I don't see why. With this PR, users of Elmish.WPF use a single standard interface to set up all logging. As an Elmish.WPF user, I would expect `WpfProgram.withLogger` to be a one-stop shop for setting up logging. Having to manually wire up "extra" loggers for the Elmish update loop adds boilerplate without utility (given that users can configure logging levels per category we define). Another point is that the fact that Elmish.WPF uses Elmish under the hood could be made an implementation detail. Not all .NET MVU frameworks use Elmish, e.g. [Fabulous](https://github.com/fsprojects/Fabulous).

Reply:
> ...the fact that Elmish.WPF uses Elmish under the hood could be made an implementation detail.

Oh, that seems right.  I will think more on this.

Reply:
Yes, I am convinced.  It is reasonable to consider that the use of Elmish by Elmish.WPF is an implement detail.  (Although we can't completely hide it because of the NuGet package dependency.)

I am fine with calling `fromElmishProgram` right away, converting and exposing `Elmish.Program` functions as `Elmish.WPF.WpfProgram` functions, and not exposing a `mapElmishProgram` function.